# Aricia agestis genome scan

Aim: set up code for genome scan between museum and modern, and modern core vs modern expanding

1. Genome scan

2. Sort according to Heliconius gff. 

3. Compare Manhattan plots

4. identify genes



## Genome scan

The samples in the raw vcf need to be renamed with shorter names. 
```
module load apps/bcftools-1.8

##see the current names
bcftools query -l 

10-AAg-19-1900-10_190312_L005_cutadapt_filtered_R1.fastq.gz
11-AAg-19-1900-11_190312_L005_cutadapt_filtered_R1.fastq.gz
12-AAg-19-1900-12_190312_L005_cutadapt_filtered_R1.fastq.gz
...

##Create a file with two columns: old_name new_name

nano C3.newnames

10-AAg-19-1900-10_190312_L005_cutadapt_filtered_R1.fastq.gz	AAg-19-1900-10_museum
11-AAg-19-1900-11_190312_L005_cutadapt_filtered_R1.fastq.gz	AAg-19-1900-11_museum
12-AAg-19-1900-12_190312_L005_cutadapt_filtered_R1.fastq.gz	AAg-19-1900-12_museum
...

bcftools reheader -s C3.newnames C3.raw.bcf > C3.raw.newnames.bcf
```

The file should be filtered correctly - minimum 18 individuals genotyped in each of the populations (museum, modernCore, modernExp)

The filtering script can be found [here](https://github.com/alexjvr1/UKButterflies/blob/master/03c_variants_filtering_ExpandingPops_July2019.sh)

