# Triplet D: some initial checks

<br>

## Contig length histogram (just for contigs with variants)

*Modern data*

![D_modern2](./images/D.modern_contigsWithVariants.png)


*Museum data*


![D_museum2](./images/D.museum_contigsWithVariants.png)

**Note:** Dashed line corresponds to contig length = 50 kb and *n* is the total number of contigs that have variants after the variant filtering.

<br>

## Contig length histogram (zoom in for contigs larger than 50kb)

*Modern data*

![D_modern3](./images/D.modern_50kbUPcontigsWithVariants.png)


*Museum data*


![D_museum3](./images/D.museum_50kbUPcontigsWithVariants.png)


**Note:** *n* is the total number of contigs larger than 50kb that have variants after the variant filtering. Similar histograms for both museum and and modern samples, but as we can see below the number of snps per contig is going to be different.  
 
 <br>

## Comparison between genome assembly and population data for the number of scaffolds >50 kb 

Species | Genome Assembly | Population data (after variant filtering)
------------ | ------------- |---------
D1 | 184 | 174
D2 | 211 | 168
D3 | 889 | 871




<br>

## Plotting number of SNPs per contig length 

*Modern data*

![D_modern1](./images/D.modern_nbSNPs_contigLength.png)


*Museum data*


![D_museum1](./images/D.museum_nbSNPs_contigLength.png)


**Note:** Red dashed line corresponds to the length of the largest contig.

<br>

>
-----------------------------------------
__*FINAL NOTE:*__

The D triplet appears to be working fine.

